node/caching-compiler.js

```diff
@@ -84,15 +86,27 @@ function readWithCache(info, content) {
   info.cache[cacheFilename] = content;
 
   if (typeof info.cacheDir === "string") {
-    // Writing cache files is something that should only happen at package
-    // development time, so it's acceptable to use fs.writeFileSync
-    // instead of some complicated asynchronous-but-atomic strategy.
-    fs.writeFileSync(absCachePath, content, "utf8");
+    scheduleWrite(absCachePath, content);
   }
 
   return content;
 }
 
+const pendingWrites = Object.create(null);
+let pendingWriteTimer;
+
+function scheduleWrite(path, content) {
+  pendingWrites[path] = content;
+  pendingWriteTimer = pendingWriteTimer || setTimeout(function () {
+    pendingWriteTimer = null;
+    Object.keys(pendingWrites).forEach(function (path) {
+      const content = pendingWrites[path];
+      delete pendingWrites[path];
+      fs.writeFileSync(path, content, "utf8");
+    });
+  }, 10);
+}
```

Comment:
This line
```js
pendingWrites[path] = content;
```
*could* technically overwrite an existing path/content entry, but I would hope the `content` would be the same, and newer content would be preferable anyway if the content was different.

My experience with writing cache files asynchronously is that they can end up corrupted if the app is killed mid-write (often an empty file is created but never written to), and that's tricky to detect when reading from the cache, since these are just `.js` files that very well could be empty.

We could potentially break up the flush, limiting it to a certain number of writes, or a certain time limit per flush, and just reschedule for 10ms later. I think that would be safe enough. Any thoughts on that?